File: java/210.course-schedule-ii.java
/*
 * @lc app=leetcode id=210 lang=java
 *
 * [210] Course Schedule II
 */

// @lc code=start
class Solution {

    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] coursesIndegree = new int[numCourses];
        // idnex is the course is courseId
        List<Integer> res = new ArrayList<>();
        Map<Integer, List<Integer>> curToNext = new HashMap<>();
        for (int[] comb : prerequisites) {
            int cur = comb[1];
            int next = comb[0];
            if (!curToNext.containsKey(cur)) {
                List<Integer> l = new ArrayList<>();
                l.add(next);
                curToNext.put(cur, l);
            } else {
                curToNext.get(cur).add(next);
            }
            coursesIndegree[next]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (coursesIndegree[i] == 0) {
                queue.offer(i);
            }
        }
        System.out.println(queue.toString());
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            res.add(cur);
            if (curToNext.containsKey(cur)) {
                for (int next : curToNext.get(cur)) {
                    coursesIndegree[next]--;
                    if (coursesIndegree[next] == 0)
                        queue.offer(next);
                }
            }
        }
        if (res.size() == numCourses) {
            int[] r = new int[numCourses];
            for (int i = 0; i < numCourses; i++) {
                r[i] = res.get(i);
            }
            return r;
        } else {
            return new int[0];
        }

    }
}
// @lc code=end

File: unsorted/js/210.course-schedule-ii.js
/*
 * @lc app=leetcode id=210 lang=javascript
 *
 * [210] Course Schedule II
 */

// @lc code=start
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
	const graph = new Map();
	const courseToindegree = new Map();
	//intialize graph course => []
	//initialize indegree  course => 0
	for (let i = 0; i < numCourses; i++) {
		courseToindegree.set(i, 0);
		graph.set(i, []);
	}
	for (const i of prerequisites) {
		const course = i[0];
		const pre = i[1];

		// do not push directly
		// const nexts = graph.get(pre);
		// nexts.push(course);

		// for array, you donot need to set....
		graph.get(pre).push(course);
		// graph.set(pre, nexts);
		//indegree
		courseToindegree.set(course, courseToindegree.get(course) + 1);
	}
	console.log('graph ', graph);
	console.log('courseToindgree ', courseToindegree);

	let inorder = [];
	let queue = [];
	// get a order from the array
	for (let [ course, indegree ] of courseToindegree.entries()) {
		if (indegree === 0) {
			queue.push(course);
		}
	}

	while (queue.length) {
		const course = queue.shift();
		inorder.push(course);
		// the course 1's indegree could be reduced
		// console.log('graph.get(course)', graph.get(course));
		for (const next of graph.get(course)) {
			const indegree = courseToindegree.get(next) - 1;
			courseToindegree.set(next, indegree);
			if (indegree === 0) {
				queue.push(next);
			}
		}
	}
	// console.log(inorder);
	if (inorder.length === numCourses) {
		return inorder;
	}
	else {
		return [];
	}
};
// @lc code=end
